dhall/src/Dhall/Core.hs

```diff
@@ -2013,9 +2013,9 @@ isNormalized e0 = loop (denote e0)
       Field r k -> case r of
           RecordLit _ -> False
           Project _ _ -> False
-          Combine x@(RecordLit m) y -> loop x && loop y && Dhall.Map.member k m
-          Combine x (RecordLit m) -> loop x && Dhall.Map.toList (fmap loop m) == [(k, True)]
-          Prefer x@(RecordLit m) y -> loop x && loop y && Dhall.Map.member k m
+          Combine (RecordLit m) _ -> Dhall.Map.member k m && loop r
```

Comment:
Perhaps move  the `loop r` check above so that does not have to be repeated across the branches, like this:

```haskell
Field r k -> loop r && case r of
    RecordLit _ -> False
    …
```

Reply:
I actually did that in a previous version of this fix, but I just couldn't handle the inefficiency! ;)

If you don't mind too much I'd slightly prefer to keep the fix as it is now.

Reply:
I don't mind!  It was just a suggestion :slightly_smiling_face: